### Importing the data for analysis:

In [1]:
import pandas as pd

df = pd.df = pd.read_csv("loan_payments.csv")

### Task 1 - Convert columns to correct format:

In [2]:
from transform_data import DataTransform

transformer = DataTransform(df)
transformer.timedelta_cols()
transformer.datetime_cols()
transformer.categorical_cols()
transformer.get_dataframe()

modified_df = transformer.get_dataframe()

print(modified_df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54231 entries, 0 to 54230
Data columns (total 43 columns):
 #   Column                       Non-Null Count  Dtype          
---  ------                       --------------  -----          
 0   id                           54231 non-null  int64          
 1   member_id                    54231 non-null  int64          
 2   loan_amount                  54231 non-null  int64          
 3   funded_amount                51224 non-null  float64        
 4   funded_amount_inv            54231 non-null  float64        
 5   term                         49459 non-null  object         
 6   int_rate                     49062 non-null  float64        
 7   instalment                   54231 non-null  float64        
 8   grade                        54231 non-null  category       
 9   sub_grade                    54231 non-null  object         
 10  employment_length            52113 non-null  object         
 11  home_ownership              

c:\Users\torig\Project_2\Customer_loans_in_finance\transform_data.py:30: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  self.df[col] = pd.to_datetime(self.df[col])
c:\Users\torig\Project_2\Customer_loans_in_finance\transform_data.py:30: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  self.df[col] = pd.to_datetime(self.df[col])
c:\Users\torig\Project_2\Customer_loans_in_finance\transform_data.py:30: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  self.df[col] = pd.to_datetime(self.df[col])
c:\Users\torig\Project_2\Customer_loans_in_finance\transform_data.py:30: UserWarning: Could

### Task 2 - Describing the data as it stands once Dtype have been updated:

In [3]:
from transform_data import DataFrameInfo

describing = DataFrameInfo(modified_df)
describing.describe()
describing.statistical_values()
describing.distinct_values()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54231 entries, 0 to 54230
Data columns (total 43 columns):
 #   Column                       Non-Null Count  Dtype          
---  ------                       --------------  -----          
 0   id                           54231 non-null  int64          
 1   member_id                    54231 non-null  int64          
 2   loan_amount                  54231 non-null  int64          
 3   funded_amount                51224 non-null  float64        
 4   funded_amount_inv            54231 non-null  float64        
 5   term                         49459 non-null  object         
 6   int_rate                     49062 non-null  float64        
 7   instalment                   54231 non-null  float64        
 8   grade                        54231 non-null  category       
 9   sub_grade                    54231 non-null  object         
 10  employment_length            52113 non-null  object         
 11  home_ownership              

{'term': 2,
 'grade': 7,
 'sub_grade': 35,
 'employment_length': 11,
 'home_ownership': 5,
 'verification_status': 3,
 'loan_status': 9,
 'payment_plan': 2,
 'purpose': 14,
 'application_type': 1}

In [4]:
describing.shape()
describing.null_counts()

 
Total number of rows: 54231
Total number of columns: 43


                             Null Count  Null Percentage
funded_amount                      3007         5.544799
term                               4772         8.799395
int_rate                           5169         9.531449
employment_length                  2118         3.905515
mths_since_last_delinq            31002        57.166565
mths_since_last_record            48050        88.602460
last_payment_date                    73         0.134609
next_payment_date                 32608        60.127971
last_credit_pull_date                 7         0.012908
collections_12_mths_ex_med           51         0.094042
mths_since_last_major_derog       46732        86.172116


### Task 3 - Remove / Impute missing values in the data 

- Reduction of data points from 54231 to 51986 
- Showing the difference in the stats once removed/imputed missing values from the data.

In [11]:
from transform_data import DataFrameTransform

transforming=DataFrameTransform(modified_df)

transforming.null_counts()
transforming.remove_high_null_columns()
transforming.impute_columns_with_median()
transforming.impute_columns_with_mode()
transforming.remove_rows_with_missing_data()
updated_df = transforming.get_updated_dataframe()

print(updated_df.info())




#Showing the changes in the stats from removing null counts:
new_stats = DataFrameInfo(updated_df)
describe_new_df = new_stats.statistical_values()
print(describe_new_df)

<class 'pandas.core.frame.DataFrame'>
Index: 51986 entries, 0 to 54227
Data columns (total 39 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   id                          51986 non-null  int64         
 1   member_id                   51986 non-null  int64         
 2   loan_amount                 51986 non-null  int64         
 3   funded_amount               51986 non-null  float64       
 4   funded_amount_inv           51986 non-null  float64       
 5   term                        51986 non-null  object        
 6   int_rate                    51986 non-null  float64       
 7   instalment                  51986 non-null  float64       
 8   grade                       51986 non-null  category      
 9   sub_grade                   51986 non-null  object        
 10  employment_length           51986 non-null  object        
 11  home_ownership              51986 non-null  category      


#### Visualiation of null values before and after transformation:

In [6]:
from transform_data import Plotter

plotting = Plotter()
plotting.plot_null_counts(df, new_df)

NameError: name 'new_df' is not defined

#### Saving it down to csv to use in milestone4

In [47]:
new_df.to_csv('C:/Users/torig/Project_2/Customer_loans_in_finance/new_dataframe.csv', index=False)

### Task 4 - Remove outliers from the data

#### Null Data: 